### PR TITLE
Add support for computing the inverse hyperbolic cotangent (acoth)

### LINKED ISSUE #236: RFC: add support for computing the inverse hyperbolic cotangent (acoth)
## Checklist

> Please ensure the following tasks are completed before submitting a feature request.

-   [x] Read and understood the [Code of Conduct][code-of-conduct].
-   [x] Searched for existing issues and pull requests.
-   [x] The issue name begins with `RFC:`.

## Description

> Description of the feature request.

This RFC proposes adding support for computing the inverse hyperbolic cotangent. The initial implementation is straightforward:

```javascript
return atanh( 1.0 / x );
```

Package: `@stdlib/math/base/special/acoth`
Alias: `base.acoth`

## Related Issues

> Does this feature request have any related issues?

No.

## Questions

> Any questions for reviewers?

No.

## Other

> Any other information relevant to this feature request? This may include screenshots, references, sample output, and/or implementation notes.

Reference implementations:

-   [Julia](https://github.com/JuliaLang/julia/blob/dee7d77c199c4776cf4e06afb7ce920f09ad3c8c/base/special/trig.jl#L983)

[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md

### PR DESCRIPTION
**Pull request by BrunoFenzl**
Resolves #236.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing], including the relevant style guides.
-   [x] Read and understand the [Code of Conduct][code-of-conduct].
-   [x] Read and understood the [licensing terms][license].
-   [x] Searched for existing issues and pull requests **before** submitting this pull request.
-   [x] Filed an issue (or an issue already existed) **prior to** submitting this pull request.
-   [x] Rebased onto latest `develop`.
-   [x] Submitted against `develop` branch.

## Description

> What is the purpose of this pull request?

This pull request:

-   implements issue #236 

## Related Issues

> Does this pull request have any related issues?
no

This pull request:

-   resolves #236

## Questions

> Any questions for reviewers of this pull request?
tex notation docs are not up to date. I have to read my way through to update them.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

Happy to have some feedback!

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md

[license]: https://github.com/stdlib-js/stdlib/blob/develop/LICENSE

### CODE REVIEW

File: lib/node_modules/@stdlib/math/base/special/acoth/README.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+-->
+
+# acoth
+
+> Compute the [inverse hyperbolic cotangent][inverse-hyperbolic-cotangent] of a number.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/README.md]:
```suggestion
> Compute the [inverse hyperbolic cotangent][inverse-hyperbolic-cotangent].
```

File: lib/node_modules/@stdlib/math/base/special/acoth/README.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+var acoth = require( '@stdlib/math/base/special/acoth' );
+```
+
+#### acoth( x )
+
+Computes the [inverse hyperbolic cotangent][inverse-hyperbolic-cotangent] of a `number` (in radians).
+
+```javascript
+var v = acoth( 1.0 );
+// returns 0
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/README.md]:
```suggestion
// returns 0.0
```

File: lib/node_modules/@stdlib/math/base/special/acoth/README.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+v = acoth( 2.0 );
+// returns ~0.5493
+
+v = acoth( NaN );
+// returns NaN
+```
+
+The domain of `x` is restricted to `[x < -1 || x > 1)`. If `x >= -1 || <= 1`, the function will return `NaN`.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/README.md]:
```suggestion
If `x` is on the interval `[-1, 1]`, the function returns `NaN`.
```

File: lib/node_modules/@stdlib/math/base/special/acoth/benchmark/benchmark.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+var bench = require( '@stdlib/bench' );
+var randu = require( '@stdlib/random/base/randu' );
+var isnan = require( '@stdlib/math/base/assert/is-nan' );
+var pkg = require( './../package.json' ).name;
+var acoth = require( './../lib' );
+
+
+// VARIABLES //
+
+var opts = {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/benchmark/benchmark.js]: This can be removed, along with the "built-in" benchmark. The built-in `Math` object does not have a `acoth` method, nor are there any proposals to add to the language.

File: lib/node_modules/@stdlib/math/base/special/acoth/benchmark/benchmark.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+// MAIN //
+
+bench( pkg, function benchmark( b ) {
+	var x;
+	var y;
+	var i;
+
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		x = ( randu()*100.0 ) + 1.0;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/benchmark/benchmark.js]:
```suggestion
		x = ( randu()*100.0 ) + 1.1;
```

`randu()` returns values on the interval `[0,1)`. In which case, need to add a value slightly greater than `1` in order to ensure that `y` is not `NaN`.

File: lib/node_modules/@stdlib/math/base/special/acoth/benchmark/benchmark.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	}
+	b.toc();
+	if ( isnan( y ) ) {
+		b.fail( 'should not return NaN' );
+	}
+	b.pass( 'benchmark finished' );
+	b.end();
+});
+
+bench( pkg+'::built-in', opts, function benchmark( b ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/benchmark/benchmark.js]: This benchmark can be removed.

File: lib/node_modules/@stdlib/math/base/special/acoth/benchmark/c/benchmark.c [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+	double elapsed;
+	double x;
+	double y;
+	double t;
+	int i;
+
+	t = tic();
+	for ( i = 0; i < ITERATIONS; i++ ) {
+		x = ( 100.0*rand_double() ) + 1.0;
+		y = acoth( x );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/benchmark/c/benchmark.c]: `acoth` is not a built-in C/C++ function. You'll need to actually provide an implementation, either as a separate function or inlined. For example,

```suggestion
y = atanh( 1.0/x );
```

File: lib/node_modules/@stdlib/math/base/special/acoth/benchmark/c/cephes/Makefile [hunk 8]
```diff
@@ -0,0 +1,113 @@
+#/
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/benchmark/c/cephes/Makefile]: Cephes does not have an `acoth` function. You can remove this file.

File: lib/node_modules/@stdlib/math/base/special/acoth/benchmark/c/cephes/benchmark.c [hunk 9]
```diff
@@ -0,0 +1,139 @@
+/**
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/benchmark/c/cephes/benchmark.c]: Cephes does not have an `acoth` function. You can remove this file.

File: lib/node_modules/@stdlib/math/base/special/acoth/benchmark/cpp/boost/Makefile [hunk 10]
```diff
@@ -0,0 +1,110 @@
+#/
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/benchmark/cpp/boost/Makefile]: To my knowledge, Boost does not have an `acoth` function. You can remove this file.

File: lib/node_modules/@stdlib/math/base/special/acoth/benchmark/cpp/boost/benchmark.cpp [hunk 11]
```diff
@@ -0,0 +1,134 @@
+/**
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/benchmark/cpp/boost/benchmark.cpp]: To my knowledge, Boost does not have an `acoth` function. You can remove this file.

File: lib/node_modules/@stdlib/math/base/special/acoth/benchmark/julia/benchmark.jl [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+* The elapsed time is in seconds.
+
+# Examples
+
+``` julia
+julia> out = benchmark();
+```
+"""
+function benchmark()
+	t = BenchmarkTools.@benchmark acoth( (100.0*rand()) + 1.0 ) samples=1e6
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/benchmark/julia/benchmark.jl]:
```suggestion
	t = BenchmarkTools.@benchmark acoth( (100.0*rand()) + 1.1 ) samples=1e6
```

File: lib/node_modules/@stdlib/math/base/special/acoth/benchmark/python/benchmark.py [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    print("  iterations: " + str(ITERATIONS))
+    print("  elapsed: " + str(elapsed))
+    print("  rate: " + str(rate))
+    print("  ...")
+
+
+def benchmark():
+    """Run the benchmark and print benchmark results."""
+    setup = "from math import acoth; from random import random;"
+    stmt = "y = acoth(100.0*random() + 1.0)"
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/benchmark/python/benchmark.py]: Python does not have an `acoth` function. You need to modify this as follows:

```suggestion
stmt = "y = atanh(1.0 / ((100.0*random() + 1.1))"
```

File: lib/node_modules/@stdlib/math/base/special/acoth/benchmark/r/benchmark.R [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+	#'
+	#'
+	#' @param iterations Number of Iterations.
+	#' @return Elapsed time in seconds.
+	#'
+	#' @examples
+	#' elapsed <- benchmark( 10000L );
+	benchmark <- function( iterations ) {
+		# Run the benchmarks:
+		results <- microbenchmark::microbenchmark( acoth( (100.0*runif(1)) + 1.0 ), times = iterations );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/benchmark/r/benchmark.R]: To my knowledge, R does not have an `acoth` function. You need to modify this to

```suggestion
results <- microbenchmark::microbenchmark( atanh( 1.0 / ( (100.0*runif(1)) + 1.1 ) ), times = iterations );
```

File: lib/node_modules/@stdlib/math/base/special/acoth/docs/repl.txt [hunk 15]
```diff
@@ -0,0 +1,28 @@
+
+{{alias}}( x )
+    Computes the inverse hyperbolic cotangent of a number.
+
+    If `x >= -1 && x <= 1`, the function returns `NaN`.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/docs/repl.txt]:
```suggestion
    If provided a value on the interval [-1, 1], the function returns `NaN`.
```

File: lib/node_modules/@stdlib/math/base/special/acoth/docs/repl.txt [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+
+    Parameters
+    ----------
+    x: number
+        Input value.
+
+    Returns
+    -------
+    y: number|out
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/docs/repl.txt]:
```suggestion
    y: number
```

File: lib/node_modules/@stdlib/math/base/special/acoth/docs/repl.txt [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+        Inverse hyperbolic cotangent (in radians).
+
+    Examples
+    --------
+    > var y = {{alias}}( 1.0 )
+    0
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/docs/repl.txt]:
```suggestion
    0.0
```

I believe, based on the documentation, that this should be `NaN`, not `0.0`.

File: lib/node_modules/@stdlib/math/base/special/acoth/lib/acoth.js [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+var isnan = require( '@stdlib/math/base/assert/is-nan' );
+var ln = require( '@stdlib/math/base/special/ln' );
+
+
+// MAIN //
+
+/**
+* Computes the inverse hyperbolic cotangent of a number.
+*
+* ## Method
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/lib/acoth.js]: You can remove the method section in its entirety. This was only applicable for `acosh`, where we needed to explain in more detail how the implementation was determined. For `acoth`, the implementation is more straightforward.

File: lib/node_modules/@stdlib/math/base/special/acoth/lib/acoth.js [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+
+	if ( x >= -1.0 && x <= 1.0 ) {
+		return NaN;
+	}
+
+	if (!isFinite(x)) {
+		return 0;
+	}
+
+	return (ln((x + 1) / x) + ln(x / (x - 1))) / 2;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/lib/acoth.js]: This implementation does not seem right to me. Is this still a work-in-progress?

File: lib/node_modules/@stdlib/math/base/special/acoth/lib/index.js [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+/**
+* Compute the hyperbolic arccosine of a number.
+*
+* @module @stdlib/math/base/special/acoth
+*
+* @example
+* var acoth = require( '@stdlib/math/base/special/acoth' );
+*
+* var v = acoth( 1.0 );
+* // returns 0
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/lib/index.js]:
```suggestion
* // returns 0.0
```

However, I believe, based on the documentation, this should be `NaN`, not `0.0`.

File: lib/node_modules/@stdlib/math/base/special/acoth/package.json [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+    "sunos",
+    "win32",
+    "windows"
+  ],
+  "keywords": [
+    "stdlib",
+    "stdmath",
+    "mathematics",
+    "math",
+    "math.acoth",
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/package.json]: This keyword can be removed.

File: lib/node_modules/@stdlib/math/base/special/acoth/package.json [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+    "inverse",
+    "cosine",
+    "cos",
+    "arc",
+    "arccosine",
+    "trig",
+    "trigonometry",
+    "radians",
+    "angle",
+    "polyfill",
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/package.json]: This can be removed, as this package does not serve as a polyfill, given that nothing at the standards level is happening in order to add built-in support for `acoth`.

File: lib/node_modules/@stdlib/math/base/special/acoth/package.json [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+    "ponyfill"
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/package.json]: This keyword can be removed for reasons similar to `polyfill`.

File: lib/node_modules/@stdlib/math/base/special/acoth/scripts/accuracy.js [hunk 24]
```diff
@@ -0,0 +1,57 @@
+/**
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/scripts/accuracy.js]: This script can be removed.

File: lib/node_modules/@stdlib/math/base/special/acoth/scripts/fixtures/julia/runner.jl [hunk 25]
```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env julia
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/scripts/fixtures/julia/runner.jl]: This file can be removed.

File: lib/node_modules/@stdlib/math/base/special/acoth/scripts/fixtures/julia/REQUIRE [hunk 26]
```diff
@@ -0,0 +1,2 @@
+julia 0.5
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/scripts/fixtures/julia/REQUIRE]: This file can be removed.

File: lib/node_modules/@stdlib/math/base/special/acoth/scripts/fixtures/julia/data.json [hunk 27]
```diff
@@ -0,0 +1 @@
...4725274726,77.3021978021978,77.35164835164835,77.4010989010989,77.45054945054945,77.5,77.54945054945055,77.5989010989011,77.64835164835165,77.6978021978022,77.74725274725274,77.7967032967033,77.84615384615384,77.8956043956044,77.94505494505495,77.99450549450549,78.04395604395604,78.0934065934066,78.14285714285714,78.1923076923077,78.24175824175825,78.29120879120879,78.34065934065934,78.39010989010988,78.43956043956044,78.48901098901099,78.53846153846153,78.58791208791209,78.63736263736264,78.68681318681318,78.73626373626374,78.78571428571429,78.83516483516483,78.88461538461539,78.93406593406593,78.98351648351648,79.03296703296704,79.08241758241758,79.13186813186813,79.18131868131869,79.23076923076923,79.28021978021978,79.32967032967034,79.37912087912088,79.42857142857143,79.47802197802197,79.52747252747253,79.57692307692308,79.62637362637362,79.67582417582418,79.72527472527473,79.77472527472527,79.82417582417582,79.87362637362638,79.92307692307692,79.97252747252747,80.02197802197803,80.07142857142857,80.12087912087912,80.17032967032966,80.21978021978022,80.26923076923077,80.31868131868131,80.36813186813187,80.41758241758242,80.46703296703296,80.51648351648352,80.56593406593407,80.6
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/scripts/fixtures/julia/data.json]: This file can be removed.

File: lib/node_modules/@stdlib/math/base/special/acoth/test/fixtures/julia/runner.jl [hunk 28]
```diff
@@ -0,0 +1,76 @@
...
+
+# Arguments
+
+* `x`: domain
+* `name::AbstractString`: output filename
+
+# Examples
+
+``` julia
+julia> x = linspace( -1000, 1000, 2001 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/test/fixtures/julia/runner.jl]:
```suggestion
julia> x = linspace( -1000.0, 1000.0, 2001 );
```

File: lib/node_modules/@stdlib/math/base/special/acoth/test/fixtures/julia/runner.jl [hunk 29]
```diff
@@ -0,0 +1,76 @@
...
+end
+
+# Get the filename:
+file = @__FILE__;
+
+# Extract the directory in which this file resides:
+dir = dirname( file );
+
+# Positive medium values:
+x = linspace( 1.0, 3.0, 500 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/test/fixtures/julia/runner.jl]:
```suggestion
x = linspace( 1.0, 3.0, 503 );
```

Avoid nice even array lengths when generating fixtures. Otherwise, one generates predictable bit patterns, which may mask implementation flaws.

File: lib/node_modules/@stdlib/math/base/special/acoth/test/fixtures/julia/runner.jl [hunk 30]
```diff
@@ -0,0 +1,76 @@
...
+gen( x, "medium_positive.json" );
+
+# Large positive values:
+x = linspace( 3.0, 28.0, 500 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/test/fixtures/julia/runner.jl]:
```suggestion
x = linspace( 3.0, 28.0, 503 );
```

File: lib/node_modules/@stdlib/math/base/special/acoth/test/fixtures/julia/runner.jl [hunk 31]
```diff
@@ -0,0 +1,76 @@
...
+gen( x, "large_positive.json" );
+
+# Larger positive values:
+x = linspace( 28.0, 100.0, 500 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/test/fixtures/julia/runner.jl]:
```suggestion
x = linspace( 28.0, 100.0, 503 );
```

File: lib/node_modules/@stdlib/math/base/special/acoth/test/fixtures/julia/runner.jl [hunk 32]
```diff
@@ -0,0 +1,76 @@
...
+gen( x, "larger_positive.json" );
+
+# Huge positive values:
+x = linspace( 1e300, 1e308, 1001 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/test/fixtures/julia/runner.jl]:
```suggestion
x = linspace( 1e300, 1e308, 1003 );
```

File: lib/node_modules/@stdlib/math/base/special/acoth/test/test.js [hunk 33]
```diff
@@ -0,0 +1,158 @@
...
+	t.end();
+});
+
+tape( 'the function returns `NaN` if provided `NaN`', function test( t ) {
+	var v = acoth( NaN );
+	t.equal( isnan( v ), true, 'returns NaN' );
+	t.end();
+});
+
+tape( 'the function returns `NaN` if provided value between `-1` and `1`', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/test/test.js]:
```suggestion
tape( 'the function returns `NaN` if provided a value between `-1` and `1`', function test( t ) {
```

Inclusive or exclusive? Not clear. The docs seem to suggest inclusive, but various example output suggests otherwise.

File: lib/node_modules/@stdlib/math/base/special/acoth/test/test.js [hunk 34]
```diff
@@ -0,0 +1,158 @@
...
+	var v;
+	var i;
+
+	for ( i = -1; i < 1e3; i++ ) {
+		v = -(randu()*1.0e6) + (1-EPS);
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/test/test.js]:
```suggestion
		v = -(randu()*1.0e6) + (1.0-EPS);
```

We include decimals for code documentation purposes when a value and/or operations involve real-valued numbers. If a value can only be integer-valued, project convention is to omit the decimal.

**Review by kgryte — CHANGES_REQUESTED** [review]
@BrunoFenzl Thanks for working on this! This PR is useful for us, as it highlights the need for us to clarify what files are needed when, particularly when it comes to testing and benchmarking math implementations. For example, several of the files which are part of `acosh` are not applicable here. Sorry about this! Thanks for your patience and perseverance!

File: lib/node_modules/@stdlib/math/base/special/acoth/test/test.js (hunk 34 above)
Comment (BrunoFenzl) [reply lib/node_modules/@stdlib/math/base/special/acoth/test/test.js]: ah! That's why... I was going to ask you why those units were being written with decimal points. Thanks for the clarification.

File: lib/node_modules/@stdlib/math/base/special/acoth/test/test.js (hunk 33 above)
Comment (BrunoFenzl) [reply lib/node_modules/@stdlib/math/base/special/acoth/test/test.js]: Should be exclusive.

File: lib/node_modules/@stdlib/math/base/special/acoth/test/fixtures/julia/runner.jl (hunk 29 above)
Comment (BrunoFenzl) [reply lib/node_modules/@stdlib/math/base/special/acoth/test/fixtures/julia/runner.jl]: Nice! Good hint.

File: lib/node_modules/@stdlib/math/base/special/acoth/lib/acoth.js (hunk 19 above)
Comment (BrunoFenzl) [reply lib/node_modules/@stdlib/math/base/special/acoth/lib/acoth.js]: I'm no mathematician so I hope I'm not utterly unfit the job! I derived this implementation from [wolframalpha.com](https://www.wolframalpha.com/input/?i=acoth(2)). One other way of writing this I found later in [wikipedia](https://en.wikipedia.org/wiki/Inverse_hyperbolic_functions#Inverse_hyperbolic_cotangent) was `(1/2) * Math.log((x + 1) / (x - 1))`. I confess I didn't benchmark the difference.
For the case `x === 1.0 || x === -1.0` the function would return Infinity, so I thought 0 would be more consistent with other functions.

**Comment by BrunoFenzl** [comment]
Wow! Thanks for the thoroughly review and sorry for the tons of inconsistencies! I'll take more care on the other functions I did, specially concerning the implementation in other languages. I have never seen such a complete testing suite (catching docs inconsistencies is awesome btw.) so I used the `acosh` folder as a template trying not to miss anything. I really hope I didn't take too much time from you.

File: lib/node_modules/@stdlib/math/base/special/acoth/README.md (hunk 3 above)
Comment (BrunoFenzl) [reply lib/node_modules/@stdlib/math/base/special/acoth/README.md]: On a second look, if `x` is exactly -1 or 1, this implementation it will return 0. NaN will be returned if `x` is less than 1 and more than -1 but not equal. This was my mistake.

File: lib/node_modules/@stdlib/math/base/special/acoth/docs/repl.txt (hunk 16 above)
Comment (BrunoFenzl) [reply lib/node_modules/@stdlib/math/base/special/acoth/docs/repl.txt]: I thought this was causing my linting error check because the function can return a number or NaN. I added it becasue the REPL documentation provides a case with multiple output values. Is `out` in this case the wrong return type or it should return just one type?

File: lib/node_modules/@stdlib/math/base/special/acoth/docs/repl.txt (hunk 17 above)
Comment (BrunoFenzl) [reply lib/node_modules/@stdlib/math/base/special/acoth/docs/repl.txt]: NaN is for values in between -1 and 1. for these exact values, the javascript implementation return Infinity and I set this to 0 so it is more consistent with other functions.

File: lib/node_modules/@stdlib/math/base/special/acoth/docs/repl.txt (hunk 16 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/acoth/docs/repl.txt]: `NaN` is a number value, despite its "not-a-number" name. More specifically, it is a particular set of bit sequences which do not have a decimal representation when read from a value stored in double-precision floating-point format. Accordingly, even if a function returns `NaN`, using the `number` type is still appropriate.

Additionally, in this case, `out` is not a value type, so would not be appropriate in any context for parameter or return value types in REPL help documentation or in JSDoc type annotations.

You are correct, some functions in the project return values of different "types", but this is not one of those cases. :)

File: lib/node_modules/@stdlib/math/base/special/acoth/docs/repl.txt (hunk 17 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/acoth/docs/repl.txt]: For clarity, set to `0` to be consistent with what functions?

File: lib/node_modules/@stdlib/math/base/special/acoth/lib/acoth.js (hunk 19 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/acoth/lib/acoth.js]: Ah. Okay. Thanks for the clarification.

As an aside, you don't have to be a mathematician to implement mathematical functions; however, I will say that it does take a while to get a "feel" for implementing numerical algorithms. But this is something which is not conferred by credentials, but more by practice.

As a first part of this journey, one thing we rarely do is implement textbook formulas. The reality is that numerical algorithms are often approximations which frequently optimize for particular value regimes. A perhaps more relevant example than `acosh` is [`atanh`](https://github.com/stdlib-js/stdlib/blob/07a705b6ce5aa3147951f737cdb8e4ee5817a991/lib/node_modules/%40stdlib/math/base/special/atanh/lib/atanh.js#L102). Notice how the Wikipedia formulas for both `atanh` and `acoth` are rather similar, but, for `atanh` (derived from FreeBSD), we are not implementing the textbook formula. Instead, we leverage `log1p` for better handling of small values and also circumvent computation entirely for tiny values.

In short, going back to the RFC, in this case, similar to, e.g., Julia, I would recommend that the implementation literally be

```javascript
var atanh = require( '@stdlib/math/base/special/atanh' );

function acoth( x ) {
    return atanh( 1.0/x );
}
```

This will hopefully resolve all the accuracy test failures in this package's unit tests and will obviate the need for any special value consideration. For example, since `atanh( 1.0 ) = +infinity`, the return value of `acoth( 1.0 )` will naturally be zero.

**Comment by kgryte** [comment]
@BrunoFenzl No worries! This is part of the process. The reality is that getting up-to-speed with all the project conventions in addition to implementing numerical algorithms can be a dizzying task (something the project needs to do a better job of addressing by better on-boarding and documentation). So thanks for putting in all the work!

**Comment by BrunoFenzl** [comment]
@kgryte I think this PR is good now. Please let me know what I could improve if you find something wrong.

**Comment by kgryte** [comment]
@BrunoFenzl Awesome! Will review shortly... :)

File: lib/node_modules/@stdlib/math/base/special/acoth/benchmark/c/cephes/Makefile (hunk 8 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/acoth/benchmark/c/cephes/Makefile]: This file still needs to be removed.

File: lib/node_modules/@stdlib/math/base/special/acoth/benchmark/c/cephes/benchmark.c (hunk 9 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/acoth/benchmark/c/cephes/benchmark.c]: This file still needs to be removed.

File: lib/node_modules/@stdlib/math/base/special/acoth/benchmark/cpp/boost/Makefile (hunk 10 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/acoth/benchmark/cpp/boost/Makefile]: This file still needs to be removed.

File: lib/node_modules/@stdlib/math/base/special/acoth/benchmark/cpp/boost/benchmark.cpp (hunk 11 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/acoth/benchmark/cpp/boost/benchmark.cpp]: This file still needs to be removed.

File: lib/node_modules/@stdlib/math/base/special/acoth/benchmark/python/benchmark.py [hunk 35]
```diff
@@ -0,0 +1,97 @@
...
+    print("  ---")
+    print("  iterations: " + str(ITERATIONS))
+    print("  elapsed: " + str(elapsed))
+    print("  rate: " + str(rate))
+    print("  ...")
+
+
+def benchmark():
+    """Run the benchmark and print benchmark results."""
+    setup = "from math import acoth; from random import random;"
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/benchmark/python/benchmark.py]:
```suggestion
    setup = "from math import atanh; from random import random;"
```

Python does not have an `acoth` function. You need to import `atanh` as this is the function used at L76.

File: lib/node_modules/@stdlib/math/base/special/acoth/lib/acoth.js [hunk 36]
```diff
@@ -0,0 +1,65 @@
...
+* @example
+* var v = acoth( 2.0 );
+* // returns ~0.5493
+*
+* @example
+* var v = acoth( NaN );
+* // returns NaN
+*/
+function acoth( x ) {
+	if (x === 0.0) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/lib/acoth.js]: Why is this special case present? The inverse hyperbolic cotangent is not defined for `0.0`.

File: lib/node_modules/@stdlib/math/base/special/acoth/scripts/accuracy.js (hunk 24 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/acoth/scripts/accuracy.js]: This file still needs to be removed.

File: lib/node_modules/@stdlib/math/base/special/acoth/scripts/fixtures/julia/data.json (hunk 27 above)
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/acoth/scripts/fixtures/julia/data.json]: This file still needs to be removed.

File: lib/node_modules/@stdlib/math/base/special/acoth/test/fixtures/julia/runner.jl (hunk 28 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/acoth/test/fixtures/julia/runner.jl]: Need to add decimals here as shown in the suggestion.

**Review by kgryte — CHANGES_REQUESTED** [review]
@BrunoFenzl Did a quick review. Most of the remaining changes I flagged concern files which are not applicable to `acoth`.

File: lib/node_modules/@stdlib/math/base/special/acoth/lib/acoth.js (hunk 36 above)
Comment (BrunoFenzl) [reply lib/node_modules/@stdlib/math/base/special/acoth/lib/acoth.js]: atanh should return 0.0 but is returning NaN. I probably missed something, but I could not figure it out why.

File: lib/node_modules/@stdlib/math/base/special/acoth/benchmark/python/benchmark.py (hunk 35 above)
Comment (BrunoFenzl) [reply lib/node_modules/@stdlib/math/base/special/acoth/benchmark/python/benchmark.py]: strange, I had this fixed... something got missing. Will push it again

File: lib/node_modules/@stdlib/math/base/special/acoth/lib/acoth.js (hunk 36 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/acoth/lib/acoth.js]: `atanh` **does** return `0.0`. To see this in action, start the stdlib REPL,

```
$ make repl
```

In the REPL,

```
> base.atanh( 0.0 )
0.0
```

For the inverse hyperbolic cotangent `acoth`, the equivalent `atanh` call is

```
> base.atanh( 1.0 / 0.0 )
NaN
```

This result is the same in Julia, Python, etc.

That the result should be `NaN` and not `0.0` can be confirmed by examining the plot of `acoth` where the domain is not defined in the region between (-1,1), as also stated in this package's documentation.

Hope this helps! :)

**Comment by kgryte** [comment]
@BrunoFenzl Just for reference, looking at the PR commit log, instead of merging from your fork into what I assume to be the cloned project repository, you want to merge from the project remote (i.e., this repo) into your fork. This will help prevent mixing your contributions with updates from the upstream repository (e.g., this PR indicates that `119` files have changed, when this PR really only concerns about `20` files). For example, assuming you have forked the project and cloned the fork to your local machine,

```
$ cd ./path/to/your/fork/stdlib
$ git checkout <your_feature_branch>
$ git remote -v
$ git remote add upstream https://github.com/stdlib-js/stdlib.git
$ git remote -v
$ git fetch upstream
$ git merge upstream/develop
```
I believe this is the workflow for syncing a feature development branch (e.g., `feature/acoth`) with changes in this repository.

Apologies if this is what you did! Something just seems a bit off with the commit log.

**Comment by kgryte** [comment]
Ah...wait...did you commit directly to a `develop` branch in your fork? That seems to be why. Would you mind porting your changes to a `feature/*` branch (see also this [guide](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md#step-2-branch))? And then reopening a new PR from that feature branch to `develop` on this repository?

Sorry about this! I should have caught this earlier! In short, because this PR is attempting to go from `develop` to `develop`, the PR's commits are being interleaved with changes in the upstream repository, which will be problematic when I "squash-and-merge" this PR.

**Comment by BrunoFenzl** [comment]
Oh! Sorry for the silly mistake! I did fetch upstream to synch with develop and forgot to switch back to my feature branch. I'll make a new pull request from a clean branch. BTW, I got rid of those errors in julia with version 0.5.2. It was no issue at all downgrading it. Since I'm on a mac, I just deleted the app, removed the preferences and installed again. Sorry for the extra work!

**Comment by kgryte** [comment]
@BrunoFenzl No worries! We need to do a better job of making contributing a smoother process. Really appreciate your persistence and continued work!
